server/node/src/main/java/org/kaaproject/kaa/server/control/service/DefaultControlService.java

```diff
@@ -1015,6 +1019,56 @@ public void onNodeAdded(OperationsNodeInfo node) {
     return resolver.getNode(entityId);
   }
 
+  /**
+   * Used when writing OperationsNodeInfo instance in logs. OperationsNodeInfo instance has
+   * ByteBuffer fields which written incorrectly in the log file. In this method ByteBuffer fields
+   * temporary overwrite to null, print logs, and after ByteBuffer fields recover on theirs previous
+   * values.
+   *
+   * @param logString is a usual string for logging. Use '{}' for output OperationsNodeInfo and
+   *                  the second '{}' is OperationsServerResolver
+   * @param node      in which find and temporary overwrite all ByteBuffer-s to null
+   * @param resolver  OperationsServerResolver instance
+   */
+  private void writeLogWithoutByteBuffer(String logString,
+                                         OperationsNodeInfo node,
+                                         OperationsServerResolver resolver) {
+    //temporary set connection info of transports to null
+    Map<TransportMetaData, Map<VersionConnectionInfoPair, ByteBuffer>> map = new HashMap<>();
+    HashMap<VersionConnectionInfoPair, ByteBuffer> innerMap;
+    for (TransportMetaData transport : node.getTransports()) {
+      innerMap = new HashMap<>();
+      for (VersionConnectionInfoPair connectionInfoPair : transport.getConnectionInfo()) {
+        // save ByteBuffer field value
+        innerMap.put(connectionInfoPair, connectionInfoPair.getConenctionInfo());
+        // temporary remove ByteBuffer field value
+        connectionInfoPair.setConenctionInfo(null);
+      }
+      map.put(transport, innerMap);
+    }
+    //temporary set public key of connection info to null
+    ByteBuffer publicKey = node.getConnectionInfo().getPublicKey();
+    node.getConnectionInfo().setPublicKey(null);
+
+    LOG.info(logString, node, resolver);
+
+    //set fields on previous values (recover fields values)
+    node.getConnectionInfo().setPublicKey(publicKey);
+    for (TransportMetaData transport : node.getTransports()) {
+      Map<VersionConnectionInfoPair, ByteBuffer> map2 = map.get(transport);
+      if (map2 != null) {
+        for (VersionConnectionInfoPair conInfo : transport.getConnectionInfo()) {
+          for (Map.Entry<VersionConnectionInfoPair, ByteBuffer> pair : map2.entrySet()) {
+            if (pair.getKey() == conInfo) {
```

Comment:
Comparing using == instead of equals !?? Furthermore, you can use just **get** method of **map** instead of verbose for loop.